db/compaction/compaction.cc

```diff
@@ -415,6 +415,52 @@ void Compaction::MarkFilesBeingCompacted(bool mark_as_compacted) {
   }
 }
 
+Slice Compaction::StartLevelSmallestInternalKey() const {
+  // lazy initialize, return if it's already calculated
+  if (!start_level_smallest_internal_.empty()) {
+    return start_level_smallest_internal_;
+  }
+
+  assert(inputs_[0].level == start_level());
+  assert(!inputs_[0].files.empty());
+  // non-L0 inputs are sorted
+  start_level_smallest_internal_ = inputs_[0].files.front()->smallest.Encode();
+  if (start_level() == 0) {
+    const InternalKeyComparator icmp =
+        column_family_data()->internal_comparator();
+    for (const auto& file : inputs_[0].files) {
+      if (icmp.Compare(start_level_smallest_internal_,
+                       file->smallest.Encode()) > 0) {
+        start_level_smallest_internal_ = file->smallest.Encode();
+      }
+    }
+  }
```

Comment:
Do we need to calculate it when calling StartLevelSmallestInternalKey()? Can we always calculate it in the beginning of the compaction?

Reply:
we try to avoid the calculation for normal case like non-level, non-bottommost, or non tiering, so only compute and cache that when at the first time using (only for `is_leveled_compaction && start_level == 0 && is_bottommost && tiering_enabled`). I will add comment for that.

Reply:
actually, we can calculate it with `smallest/largest_user_key`. will update that.